rhythm/calendardelay.go

```diff
@@ -0,0 +1,57 @@
+/*
+Copyright (c) 2024 Diagrid Inc.
+Licensed under the MIT License.
+*/
+
+package rhythm
+
+import "time"
+
+// CalendarDelaySchedule represents a recurring duty cycle that includes days, months or years too, e.g. "Every 3 months".
+// It does not support jobs more frequent than once a second.
+type CalendarDelaySchedule struct {
+	years  int
+	months int
+	days   int
+
+	Delay time.Duration
+}
+
+func EveryCalendar(years, months, days int, duration time.Duration) CalendarDelaySchedule {
+	if (duration < time.Second) && (years == 0) && (months == 0) && (days == 0) {
+		// Second is the minimum granularity we support.
+		duration = time.Second
+	}
+
+	return CalendarDelaySchedule{
+		years:  years,
+		months: months,
+		days:   days,
+		Delay:  duration.Truncate(time.Second),
+	}
+}
+
+// Next returns the next time this should be run.
+// This rounds so that the next activation time will be on the second.
+func (schedule CalendarDelaySchedule) Next(start, t time.Time) time.Time {
+	if start.IsZero() {
+		// schedule is not bound to a starting point
+		return t.Truncate(time.Second).Add(schedule.Delay).AddDate(schedule.years, schedule.months, schedule.days)
```

Comment:
Why truncate?

Reply:
We want to keep the second granularity and avoid the triggers to slowly skip one due to small delays between triggers.

Reply:
How could a trigger ever be skipped if we did not truncate?

Reply:
Trigger 1 is due at 8:00:00.000, but the next time this runs, the time given is 8:00:00.501, so trigger 2 would be delayed
Trigger 2 is due at 8:00:01.501, but the next time this runs, the time given is 8:00:02.003, so trigger 3 would be delayed
Trigger 3 is due at 8:00:03:002, which means we skipped 1s trigger.

By looking at the code again, the scheduled time is what is passed back and not the trigger time, so we don't risk skipping.

On the other hand, it is what the cron logic uses to trigger multiple jobs at the same second - which is the granularity we support. Otherwise, it would be too many memory operations (sorting the array) per tick.